src/server/campaignd/server.cpp

```diff
@@ -668,29 +753,110 @@ void server::handle_request_campaign_list(const server::request& req)
 
 void server::handle_request_campaign(const server::request& req)
 {
-	config& campaign = get_campaign(req.cfg["name"]);
+	config& campaign = get_addon(req.cfg["name"]);
 
 	if(!campaign || campaign["hidden"].to_bool()) {
 		send_error("Add-on '" + req.cfg["name"].str() + "' not found.", req.sock);
 		return;
 	}
-	
-	const int size = filesystem::file_size(campaign["filename"]);
 
-	if(size < 0) {
-		send_error("Add-on '" + req.cfg["name"].str() + "' could not be read by the server.", req.sock);
-		return;
+	// The desired version is selected on the client side considering
+	// the min_wesnoth_version in [version] children of the addon info (#TODO: gfgtdf's part)
+	std::string to = req.cfg["version"].str();
+	const std::string& from = req.cfg["from_version"].str();
+	std::string full_pack = campaign["filename"].str();
+
+	auto version_map = get_version_map(campaign);
+
+	if(version_map.empty()) {
+		// Old format ?
+		ERR_CS << "The (" + to + ") version of the addon '" << req.cfg["name"].str()
+			   << "' is stored in the old format (bug?)! Trying to send a legacy full-pack.";
+		to = campaign["version"].str();
+	} else {
+		auto version = version_map.find(version_info(to));
+		if(version != version_map.end()) {
+			full_pack += version->second["filename"].str();
+		} else {
+			WRN_CS << "The selected version (" << to << ") for the addon '" << req.cfg["name"].str()
+				   << "' has not been found or it is unspecified! Sending the latest version instead!\n";
+			to = version_map.rbegin()->first;
+			full_pack += version_map.rbegin()->second["filename"].str();
+		}
+
+		// Negotiate an update pack if possible
+		if(!from.empty() && version_map.count(version_info(from)) != 0) {
+			config pack_data;
+			// Make a line of consequential updates beginning from the old version to the new one.
+			// Every pair of increasing versions on the server side should contain an update_pack
+			// transition guaranteed during the upload.
+
+			auto iter = version_map.begin();
+			int size = 0;
+
+			while(std::distance(iter, version_map.end()) > 1) {
+				const config& prev_version = iter->second;
+				iter++;
+				const config& next_version = iter->second;
+
+				for(const config& pack : campaign.child_range("update_pack")) {
+					if(pack["from"].str() == prev_version["version"].str()
+							&& pack["to"].str() == next_version["version"].str()) {
+						config update_pack;
+						filesystem::scoped_istream in = filesystem::istream_file(campaign["filename"].str() + pack["filename"].str());
+						read_gz(update_pack, *in);
+						if(update_pack) {
+							pack_data.append(update_pack);
+							size += filesystem::file_size(campaign["filename"].str() + pack["filename"].str());
+						} else {
+							WRN_CS << "Unable to create an update pack sequence from version (" << from << ") to ("
+								   << to << ") for the addon '" << req.cfg["name"].str() << "'. A full pack will be sent instead!\n";
+							pack_data = config::get_invalid();
```

Comment:
This looks  wrong, assigning the invalid config to a config with just raise an error, also that break will just scape form the  for loop not form the outer while  loop

Reply:
Ok, I'll test it and fix the loop with a check in the end/start, probably.